File: monorepo_kotlin/apps/web/src/main/kotlin/mono/web/Cycle.kt
package mono.web

// Kotlin class-class cycle — same package, no special handling needed.

class CycleA(val name: String) {
    var child: CycleB? = null                 // C1.a.type_ref → CycleB

    fun spawnBravo(): CycleB = CycleB("$name/b")
    fun describe(): String    = "CycleA($name)"

    companion object {
        fun kickOff(): String {
            val a = CycleA("root")
            val b = a.spawnBravo()
            return b.bounceToAlpha()
        }
    }
}

class CycleB(val tag: String) {
    var owner: CycleA? = null                 // C1.b.type_ref → CycleA

    fun bounceToAlpha(): String = CycleA("bounce-from-$tag").describe()
}

File: swift/cycle.swift
// Swift class cycle — use `weak` on one side to avoid a retain cycle
// at runtime (ARC would otherwise leak). The LSP must still resolve
// both directions symbolically.

class AlphaS {
    let name: String
    var child: BravoS?              // strong → BravoS
    init(name: String) { self.name = name }
    func describe() -> String { return "AlphaS(\(name))" }
    func spawnBravo() -> BravoS {
        let b = BravoS(tag: "\(name)/b")
        b.owner = self              // weak back-ref
        child = b
        return b
    }
}

class BravoS {
    let tag: String
    weak var owner: AlphaS?         // weak → breaks cycle at runtime
    init(tag: String) { self.tag = tag }
    func bounceToAlpha() -> String {
        return AlphaS(name: "bounce-from-\(tag)").describe()
    }
}

func runCycleDemoSwift() {
    let a = AlphaS(name: "root")
    let b = a.spawnBravo()
    print("cycle:", b.bounceToAlpha())
}
